iceoryx_utils/include/iceoryx_utils/internal/concurrent/dual_access_transaction_tray.hpp

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) 2021 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_UTILS_CONCURRENT_DUAL_ACCESS_TRANSACTION_TRAY_HPP
+#define IOX_UTILS_CONCURRENT_DUAL_ACCESS_TRANSACTION_TRAY_HPP
+
+#include "iceoryx_utils/posix_wrapper/semaphore.hpp"
+
+#include <atomic>
+
+namespace iox
+{
+namespace concurrent
+{
+class DualAccessTransactionTray
+{
+  private:
+    /// @note not public to prevent the user from passing AccessToken::NONE to a method
+    enum class AccessToken
+    {
+        NONE,
+        LEFT,
+        RIGHT
+    };
+
+  public:
+    static constexpr AccessToken LEFT{AccessToken::LEFT};
+    static constexpr AccessToken RIGHT{AccessToken::RIGHT};
+
+    class AccessGuard
```

Comment:
I modeled this after the `std::lock_guard` but I'm not really convinced about this. Currently I'm leaning more towards something like
```c++
DualAccessTransactionTray transactionTray;
transactionTray.acquireExclusiveAccess(DualAccessTransactionTray::LEFT, [&] {
    // do stuff
});
```